server/src/main/java/org/elasticsearch/index/mapper/MapperService.java

```diff
@@ -352,24 +355,73 @@ public void merge(IndexMetadata indexMetadata, MergeReason reason) {
         }
     }
 
-    public DocumentMapper merge(String type, CompressedXContent mappingSource, MergeReason reason) {
+    public DocumentMapper merge(String type, List<CompressedXContent> mappingSources, MergeReason reason) {
+        Explicit<Boolean> subobjects = null;
+
         final DocumentMapper currentMapper = this.mapper;
-        if (currentMapper != null && currentMapper.mappingSource().equals(mappingSource)) {
-            return currentMapper;
+        if (currentMapper != null) {
+            if (mappingSources.size() == 1 && currentMapper.mappingSource().equals(mappingSources.get(0))) {
+                return currentMapper;
+            }
+            if (currentMapper.mapping().getRoot().subobjectsExplicit()) {
+                subobjects = Explicit.explicitBoolean(currentMapper.mapping().getRoot().isEnabled());
+            }
```

Comment:
We probably don't need this as the current mapper already has the correct `subobjects` setting.